src/main/ruby/truffleruby/core/integer.rb

```diff
@@ -350,7 +350,17 @@ def self.try_convert(obj)
   def self.sqrt(n)
     n = Primitive.rb_to_int(n)
     raise Math::DomainError if n.negative?
-    Math.sqrt(n).floor
+    return Math.sqrt(n).floor if n < 0xfffffffffffff
+
+    shift = n.bit_length / 4
+    x = sqrt(n >> (2 * shift)) << shift
+    x = (x + n / x) / 2
+    xx = x * x
+    while xx > n
+      xx -= 2 * x - 1
+      x -= 1
+    end
+    x
```

Comment:
Wondering how the constant `0xfffffffffffff` was chosen. Is it based on benchmarking results?